### ISSUE TITLE
Redirects not applied on subsequent runs of the push

### ISSUE THREAD

**Issue by artemgoncharuk** [laurilehmijoki/s3_website#176]
Hi,

Cannot figure out what I'm doing wrong with redirects. They are applied OK on the first run of the push, but erased on the subsequent runs:

Config has:
redirects:
  test.html: /target/

Output of the 1st run:
[succ] Redirected test.html to /target
[succ] Created test.html (max-age=86400 | text/html; charset=utf-8 | gzip)

The redirect is applied correctly on the first run of the push and "Website Redirect Location" is set correctly in the resource metadata in S3.

The subsequent run of the push, however, removed the redirect.
Output of the 2nd run:
[succ] Redirected test.html to /target
[succ] Updated test.html (max-age=86400 | text/html; charset=utf-8 | gzip)

Now the metadata does not have "Website Redirect Location"....

Any ideas how to fix this?

Thanks!

**Comment by artemgoncharuk** [comment]
Hi, 

Does anybody know a solution for this?

Thanks!

**Comment by laurilehmijoki** [comment]
This might be a bug in s3_website. Can someone try to find the root cause by debugging? Unfortunately I have no time to dedicate on this project at the moment.

**Comment by laurilehmijoki** [comment]
@artemgoncharuk I could not reproduce this bug. Unfortunately I don't have any good explanations on the problem you are experiencing.

**Comment by IWantPizza** [comment]
@artemgoncharuk Did you ever solve this? And are/were you using CloudFront?

**Comment by artemgoncharuk** [comment]
Yes, I think, the issue was in the fact that i had the following setup.

There was a file foo.html that was published to S3.
At the same time, a redirect was applied to it.

Redirects assume that there would not be a real file - it creates a 0 length file in S3 and next applies redirects. Most likely, the file creation fails (since it already exists), and due to incorrect error handling, redirect is not applied. That's my guess anyway. I have not had time to look at the code yet.

W/o the file, everything works well.

**Comment by laurilehmijoki** [comment]
Will `s3_website push --force` help you?

**Comment by vijayabaskar** [comment]
@laurilehmijoki ,  s3_website push --force  doesn't work.  It behaves the same.  It doesn't change the redirect.  @artemgoncharuk is correct, it works fine if there is no file exist with the redirect key.  Do you think there will be a fix for this?

thanks!

**Comment by laurilehmijoki** [comment]
@vijayabaskar I agree that the `--force` option should replace the existing file with the redirect setting.

**Comment by laurilehmijoki** [comment]
@vijayabaskar now that I think of this issue, I'm quite sure that the `--force` option is not a good way to express the intent "I wish to override the existing file with a redirect". It's not a good way, because if one would first push with the `--force` option (the file would be deleted and the redirect applied) and then push _without_ the option, the redirect would be replaced with the file. This is probably not what you would prefer.

I'm not sure what would be a good way to solve this issue – any suggestions?

**Comment by vijayabaskar** [comment]
@laurilehmijoki agree with you.  Is it possible to just add the redirect meta-data if the redirect is specified for non-zero file?  and add a zero length file with meta-data if the file doesn't exist?

thanks.

**Comment by laurilehmijoki** [comment]
Sounds like a good plan, @vijaybaskar. At the moment I don't see any problems in the idea. 

Due to my other commitments I can't give an estimate for the release date.
